pkg/listwatch/listwatch.go

```diff
@@ -0,0 +1,329 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Provides multiple namespace listerWatcher. This implementation is Largely from
+// https://github.com/coreos/prometheus-operator/pkg/listwatch/listwatch.go
+
+package listwatch
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"istio.io/istio/pkg/servicemesh/controller"
+	"istio.io/pkg/log"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+type NamespaceListerWatcher interface {
+	cache.ListerWatcher
+	controller.MemberRollListener
+}
+
+// MultiNamespaceListerWatcher takes a list of namespaces and a
+// cache.ListerWatcher generator func and returns a single cache.ListerWatcher
+// capable of operating on multiple namespaces.
+func MultiNamespaceListerWatcher(namespaces []string, f func(string) cache.ListerWatcher) NamespaceListerWatcher {
+	return newMultiListerWatcher(namespaces, f)
+}
+
+// multiListerWatcher abstracts several cache.ListerWatchers, allowing them
+// to be treated as a single cache.ListerWatcher.
+
+type listerWatcherState int
+
+const (
+	created      listerWatcherState = 0
+	listed       listerWatcherState = 1
+	watching     listerWatcherState = 2
+	errorOnWatch listerWatcherState = 3
+	stopping     listerWatcherState = 4
+)
+
+type listerWatcher struct {
+	lw      cache.ListerWatcher
+	stopper func()
+}
+
+type multiListerWatcher struct {
+	namespaces []string
+	f          func(string) cache.ListerWatcher
+	state      listerWatcherState
+	result     chan watch.Event
+	stopped    chan struct{}
+	wg         *sync.WaitGroup
+	lwMap      map[string]*listerWatcher
+	lock       sync.Mutex
+}
+
+type combinedEvent struct {
+	namespace string
+	event     watch.Event
+}
+
+func newMultiListerWatcher(namespaces []string, f func(string) cache.ListerWatcher) *multiListerWatcher {
+	return &multiListerWatcher{
+		namespaces: append(namespaces[:0:0], namespaces...),
```

Comment:
what is `namespaces[:0:0]` doing?

Reply:
It is basically giving you a fresh slice of the same type as `namespaces`; it not just has length `0` but also capacity `0`, the second part making sure you're not pointing to the same memory. Another way would be defining a new `[]string` and `append`ing to that.

Reply:
Note that this comment has since disappeared because the change was coming in from the already-merged MultiListerWatcher PR

Reply:
Thanks. Odd syntax for that.